src/toast/todmap/sim_det_map.py

```diff
@@ -166,13 +189,51 @@ def exec(self, data):
             None
 
         """
+        log = Logger.get()
+
+        if isinstance(self._input_map, DistPixels):
+            input_map = self._input_map
+        elif "DETECTOR" not in self._input_map:
+            if data.comm is None or data.comm.world_rank == 0:
+                log.info("Scanning {}".format(self._input_map))
+                if not os.path.isfile(self._input_map):
+                    raise RuntimeError(
+                        "Input map not found: {}".format(self._input_map)
+                    )
+            input_map = DistPixels(
+                data, nnz=self._nnz, dtype=np.float32, pixels=self._pixels
+            )
+            input_map.read_healpix_fits(self._input_map)
+        else:
+            input_map = None
 
         for obs in data.obs:
             tod = obs["tod"]
 
-            dets = tod.local_dets if self._dets is None else self._dets
+            if self._dets is None:
+                dets = tod.local_dets
+            else:
+                dets = self._dets
 
             for det in dets:
+                if input_map is None:
+                    if MPI is None:
+                        comm = None
+                    else:
+                        comm = MPI.COMM_SELF
+                    filename = self._input_map.replace("DETECTOR", det)
+                    if not os.path.isfile(filename):
+                        raise RuntimeError("Input map not found: {}".format(filename))
+                    detector_map = DistPixels(
+                        data,
+                        comm=comm,
+                        nnz=self._nnz,
+                        dtype=np.float32,
+                        pixels=self._pixels,
+                    )
+                    detector_map.read_healpix_fits(filename)
```

Comment:
You are looping over local detectors (potentially different on every process), but distributing the map over the communicator used for the whole observation.  If each process had the same number of detectors (loop iterations), then the filename on rank zero would be loaded into detector_map regardless of what detectors existed on other processes.  In the case of different numbers of detectors on each process you would have a hang on some processes waiting for communication.

It seems like some options would be:

1. Loop over the full detector list and only scan the map to timestreams if the detector is locally present on the process.
2.  OR, Loop over local detectors and use the process grid row communicator for the DistPixels object.

The second option would result in each row of the process grid independently scanning their detector maps.

Reply:
I don't think that to be the case. Lines 221-224 and 230 are there to ensure that every process loads and samples the map independent of other processes. I didn't want to use the rank communicators because we can't guarantee that same detectors are there for every rank.

Reply:
Ok, sorry I missed that- thanks for correcting my understanding.  In the case of scanning a single map for all detectors, that is done once (collectively) at the beginning of the function and then used for all observations.  In the case of detector-specific maps this code is basically doing:
```
for each observation:
    for each local detector:
        each process reads detector map from a file
            - DistPixels has the locally hit submaps for all detectors
              and all observations on this process 
        project to timestream
        discard map
```
So that means if you have 100 observations then a detector map gets loaded off disk 100 times for every process that has that detector.  I guess we are not running in this mode very much currently, but in the future we could consider each process doing a loop over observations to compute the superset of all detectors and then loop over detectors and populate the observations where it has that detector.  However, I am fine with this implementation if it serves our needs for now.

Reply:
Yes, this is not optimal but it is a relatively clean solution that requires no blocking communication. Also, parallel reads tend to be efficient, especially when the files are only accessed by less than a hundred processes.